File: src/main/java/org/micoli/php/exportSourceToMarkdown/ExportSourceToMarkdownService.java
package org.micoli.php.exportSourceToMarkdown;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.knuddels.jtokkit.Encodings;
import com.knuddels.jtokkit.api.Encoding;
import com.knuddels.jtokkit.api.EncodingRegistry;
import com.knuddels.jtokkit.api.EncodingType;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.micoli.php.attributeNavigation.service.FileData;
import org.micoli.php.exportSourceToMarkdown.configuration.ExportSourceToMarkdownConfiguration;
import org.micoli.php.service.filesystem.FileListProcessor;
import org.micoli.php.ui.Notification;
import org.thymeleaf.TemplateEngine;
import org.thymeleaf.context.Context;
import org.thymeleaf.templateresolver.StringTemplateResolver;

public class ExportSourceToMarkdownService {
    private final Project project;

    public ExportSourceToMarkdownService(Project project) {
        this.project = project;
    }

    public static ExportSourceToMarkdownService getInstance(Project project) {
        return project.getService(ExportSourceToMarkdownService.class);
    }

    private ExportSourceToMarkdownConfiguration configuration = new ExportSourceToMarkdownConfiguration();

    public void loadConfiguration(ExportSourceToMarkdownConfiguration exportSourceToMarkdownConfiguration) {
        if (exportSourceToMarkdownConfiguration == null) {
            return;
        }
        configuration = exportSourceToMarkdownConfiguration;
    }

    public ExportedSource generateMarkdownExport(VirtualFile[] selectedFiles) {

        List<VirtualFile> fileList = FileListProcessor.findFilesFromSelectedFiles(List.of(selectedFiles));
        if (fileList.isEmpty()) {
            return null;
        }

        ContextualAmender contextualAmender = new ContextualAmender(project, configuration);

        List<VirtualFile> filesInContext =
                getUseContextualNamespaces() ? contextualAmender.amendListWithContextualFiles(fileList) : fileList;
        VirtualFile ignoreFile = getIgnoreFile();
        List<VirtualFile> filteredFiles = FileListProcessor.filterFiles(ignoreFile, filesInContext);

        Context context = new Context();
        context.setVariable("files", getFileData(sortFiles(filteredFiles)));

        String exportContent = getTemplateEngine().process(configuration.template, context);

        return new ExportedSource(exportContent, getNumberOfTokens(exportContent));
    }

    private @Nullable VirtualFile getIgnoreFile() {
        if (getUseIgnoreFile()) {
            VirtualFile virtualFile = ProjectUtil.guessProjectDir(project);
            if (virtualFile != null) {
                return virtualFile.findChild(".aiignore");
            }
        }
        return null;
    }

    private @NotNull List<VirtualFile> sortFiles(List<VirtualFile> filesInContext) {
        return new ArrayList<>(Set.copyOf(filesInContext))
                .stream()
                        .sorted(Comparator.comparing(VirtualFile::getPath).thenComparing(VirtualFile::getName))
                        .toList();
    }

    private @NotNull List<FileData> getFileData(List<VirtualFile> processedFiles) {
        List<FileData> files = new ArrayList<>();
        String baseDir = project.getBasePath();
        assert baseDir != null;

        for (VirtualFile file : processedFiles) {
            try {
                String content = new String(file.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
                String extension = file.getExtension() != null ? file.getExtension() : "plain";
                files.add(new FileData(file.getPath().replace(baseDir, ""), content, extension));
            } catch (IOException e) {
                Notification.getInstance(project).error(e.getMessage());
            }
        }
        return files;
    }

    private @NotNull TemplateEngine getTemplateEngine() {
        TemplateEngine templateEngine = new TemplateEngine();
        StringTemplateResolver resolver = new StringTemplateResolver();
        resolver.setTemplateMode("TEXT");
        templateEngine.setTemplateResolver(resolver);
        return templateEngine;
    }

    private int getNumberOfTokens(String exportContent) {
        EncodingRegistry registry = Encodings.newDefaultEncodingRegistry();
        Encoding enc = registry.getEncoding(EncodingType.CL100K_BASE);

        return enc.countTokens(exportContent);
    }

    public void toggleUseContextualNamespaces() {
        configuration.useContextualNamespaces = !configuration.useContextualNamespaces;
    }

    public boolean getUseContextualNamespaces() {
        return configuration.useContextualNamespaces;
    }

    public boolean getUseIgnoreFile() {
        return configuration.useIgnoreFile;
    }

    public void toggleUseIgnoreFile() {
        configuration.useIgnoreFile = !configuration.useIgnoreFile;
    }
}

File: src/main/kotlin/org/micoli/php/exportSourceToMarkdown/ExportSourceToMarkdownService.kt
package org.micoli.php.exportSourceToMarkdown

import com.intellij.openapi.components.Service
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.vfs.VirtualFile
import com.knuddels.jtokkit.Encodings
import com.knuddels.jtokkit.api.EncodingType
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.Set
import org.micoli.php.attributeNavigation.service.FileData
import org.micoli.php.exportSourceToMarkdown.configuration.ExportSourceToMarkdownConfiguration
import org.micoli.php.service.filesystem.FileListProcessor
import org.micoli.php.service.filesystem.FileListProcessor.findFilesFromSelectedFiles
import org.micoli.php.ui.Notification
import org.thymeleaf.TemplateEngine
import org.thymeleaf.context.Context
import org.thymeleaf.templateresolver.StringTemplateResolver

@Service(Service.Level.PROJECT)
class ExportSourceToMarkdownService(private val project: Project) {
    private var configuration: ExportSourceToMarkdownConfiguration =
        ExportSourceToMarkdownConfiguration()

    fun loadConfiguration(
        exportSourceToMarkdownConfiguration: ExportSourceToMarkdownConfiguration?
    ) {
        if (exportSourceToMarkdownConfiguration == null) {
            return
        }
        configuration = exportSourceToMarkdownConfiguration
    }

    fun generateMarkdownExport(selectedFiles: Array<VirtualFile>): ExportedSource? {
        val fileList = findFilesFromSelectedFiles(mutableListOf(*selectedFiles))
        if (fileList.isEmpty()) {
            return null
        }

        val contextualAmender = ContextualAmender(project, configuration)

        val filesInContext =
            if (this.useContextualNamespaces)
                contextualAmender.amendListWithContextualFiles(fileList)
            else fileList
        val ignoreFile = this.ignoreFile
        val filteredFiles = FileListProcessor.filterFiles(ignoreFile, filesInContext)

        val context = Context()
        context.setVariable("files", getFileData(sortFiles(filteredFiles)))

        val exportContent = this.templateEngine.process(configuration.template, context)

        return ExportedSource(exportContent, getNumberOfTokens(exportContent))
    }

    private val ignoreFile: VirtualFile?
        get() {
            if (this.useIgnoreFile) {
                val virtualFile = project.guessProjectDir()
                if (virtualFile != null) {
                    return virtualFile.findChild(".aiignore")
                }
            }
            return null
        }

    private fun sortFiles(filesInContext: MutableList<VirtualFile>): MutableList<VirtualFile> {
        return Set.copyOf<VirtualFile>(filesInContext)
            .stream()
            .sorted(
                Comparator.comparing<VirtualFile, String> { it.path }
                    .thenComparing<String> { it.name })
            .toList()
    }

    private fun getFileData(processedFiles: MutableList<VirtualFile>): MutableList<FileData> {
        val files = ArrayList<FileData>()
        val baseDir = checkNotNull(project.basePath)
        for (file in processedFiles) {
            try {
                val content = String(file.inputStream.readAllBytes(), StandardCharsets.UTF_8)
                val extension = file.extension ?: "plain"
                files.add(FileData(file.path.replace(baseDir, ""), content, extension))
            } catch (e: IOException) {
                Notification.getInstance(project).error(e.localizedMessage)
            }
        }
        return files
    }

    private val templateEngine: TemplateEngine
        get() {
            val templateEngine = TemplateEngine()
            val resolver = StringTemplateResolver()
            resolver.setTemplateMode("TEXT")
            templateEngine.setTemplateResolver(resolver)
            return templateEngine
        }

    private fun getNumberOfTokens(exportContent: String?): Int {
        val registry = Encodings.newDefaultEncodingRegistry()
        val enc = registry.getEncoding(EncodingType.CL100K_BASE)

        return enc.countTokens(exportContent)
    }

    fun toggleUseContextualNamespaces() {
        configuration.useContextualNamespaces = !configuration.useContextualNamespaces
    }

    val useContextualNamespaces: Boolean
        get() = configuration.useContextualNamespaces

    val useIgnoreFile: Boolean
        get() = configuration.useIgnoreFile

    fun toggleUseIgnoreFile() {
        configuration.useIgnoreFile = !configuration.useIgnoreFile
    }

    companion object {
        @JvmStatic
        fun getInstance(project: Project): ExportSourceToMarkdownService {
            return project.getService(ExportSourceToMarkdownService::class.java)
        }
    }
}
